### PR TITLE
Set the global function generating aggregate identifiers

### PR DESCRIPTION
**Pull request by hallgren**
First, sorry for poking around in your code. I'm just interested how you are using my set of event sourcing packages. This is the first implementation that is using the projections part. 🥇 

Please feel free to give me any feedback on your experience using it.

I created this small PR showing how you could set the global SetID function instead of having to remember setting the identifier on all newly created aggregates.

BR Morgan

### CODE REVIEW

**Comment by ehsanranjbar** [comment]
Hey! First of all we are very glad that our codebase piqued your interest. and I'm grateful for the PR. We are doing this project for a Hackathon so I didn't have enough time to learn about all the features of your library. 😅
I'm a DDD newbie but coming from Rust and [cqrs-es](https://docs.rs/cqrs-es/latest/cqrs_es/) to Go and using your package  was very straight forward.

Regarding the experience with `Projection`s I think there's some stuff that can be improved.
So first of all there's (IMO) this problem with `Group` where it doesn't run a projection again when it returns error. In our case because we are using SQLite which has a limitation on accessing tables concurrently apparently (`database is locked` error) some of our `Projection`s fail and never gets to try again so our read models get stuck.
Second i think it's overall better to have a mechanism to broadcast events to `Projection`s as they are getting saved by `EventRepository` instead of reading the events manually. I mean in the large scale there might be a separate queue or database for events (e.g. Kafka) and it makes sense for `Projection`s to run independently and fetch events on their own but in our small scales events, read and write models all reside in one application and database so it's more reasonable to have a strategy for Projections to get events in real-time like the `Subscribe` and only access the database when it falls behind.

Also with the introduction of generics in Go, i think there are some areas that can be improved with generics. There were some other event-sourcing packages for Go that used generics but yours was the most reliable at the time.

File: app/app.go [hunk 1]
```diff
@@ -23,6 +24,9 @@ type Application struct {
 }
 
 func NewApplication(db *sql.DB) (*Application, error) {
+	// Set how identifiers are generated on newly created aggregates
+	eventsourcing.SetIDFunc(uuid.New().String)
```
Comment (ehsanranjbar) [inline app/app.go]:
```suggestion
	eventsourcing.SetIDFunc(func() string {
		return uuid.New().String()
	})
```

**Review by ehsanranjbar — CHANGES_REQUESTED** [review]

**Comment by hallgren** [comment]
> Hey! First of all we are very glad that our codebase piqued your interest. and I'm grateful for the PR. We are doing this project for a Hackathon so I didn't have enough time to learn about all the features of your library. 😅 I'm a DDD newbie but coming from Rust and [cqrs-es](https://docs.rs/cqrs-es/latest/cqrs_es/) to Go and using your package was very straight forward.

I'm just happy to help and try to understand how the set of packages are used :) I've been trying to learn Rust but have failed multiple times. 

> 
> Regarding the experience with `Projection`s I think there's some stuff that can be improved. So first of all there's (IMO) this problem with `Group` where it doesn't run a projection again when it returns error. In our case because we are using SQLite which has a limitation on accessing tables concurrently apparently (`database is locked` error) some of our `Projection`s fail and never gets to try again so our read models get stuck.

One idea is to handle the `database is locked` error in the `Callback` function. Currently an error in any projection updates are returned. (Not sure the `database is locked` error can be handled by a retry for example.)

At work we use sqlite and when starting a write transaction we have a `sync.Mutex` to make sure there is only one write transaction.

>Second i think it's overall better to have a mechanism to broadcast events to `Projection`s as they are getting saved by `EventRepository` instead of reading the events manually. I mean in the large scale there might be a separate queue or database for events (e.g. Kafka) and it makes sense for `Projection`s to run independently and fetch events on their own but in our small scales events, read and write models all reside in one application and database so it's more reasonable to have a strategy for Projections to get events in real-time like the `Subscribe` and only access the database when it falls behind.

There is a real-time event subscription concept https://github.com/hallgren/eventsourcing?tab=readme-ov-file#event-subscription. I have used it for signalling within the application and not for building projections. But it could be okey for a proof of concept application? 

> Also with the introduction of generics in Go, i think there are some areas that can be improved with generics. There were some other event-sourcing packages for Go that used generics but yours was the most reliable at the time.

I have not dig deep into generics, but I would be interested if you have some examples or ideas you could share.

BR Morgan

**Comment by hallgren** [comment]
Hi @ehsanranjbar, I have experimented with a way of triggering projections manually. https://github.com/hallgren/eventsourcing/pull/124

`TriggerAsync`: Trigger and return.
`TriggerSync`: Trigger and wait for the projection to finish.

The methods are both on the projection and the group of projections.

Would it help in your case?

BR Morgan

**Pull request merged**